bson/bsoncodec/struct_codec.go

```diff
@@ -484,38 +518,67 @@ func (sc *StructCodec) describeStruct(r *Registry, t reflect.Type) (*structDescr
 					return nil, err
 				}
 				for _, fd := range inlinesf.fl {
-					if _, exists := sd.fm[fd.name]; exists {
-						return nil, fmt.Errorf("(struct %s) duplicated key %s", t.String(), fd.name)
-					}
 					if fd.inline == nil {
 						fd.inline = []int{i, fd.idx}
 					} else {
 						fd.inline = append([]int{i}, fd.inline...)
 					}
+					if old, exists := sd.fm[fd.name]; exists {
+						compared := compareField(old, fd)
+						if compared == 0 || !sc.AllowOverwritingInlinedFields {
+							return nil, fmt.Errorf("struct %s) duplicated key %s", t.String(), fd.name)
+						}
+						if compared > 0 {
+							continue
+						}
+					}
 					sd.fm[fd.name] = fd
-					sd.fl = append(sd.fl, fd)
+
 				}
 			default:
 				return nil, fmt.Errorf("(struct %s) inline fields must be a struct, a struct pointer, or a map", t.String())
 			}
 			continue
 		}
 
-		if _, exists := sd.fm[description.name]; exists {
-			return nil, fmt.Errorf("struct %s) duplicated key %s", t.String(), description.name)
+		if old, exists := sd.fm[description.name]; exists {
+			compared := compareField(old, description)
+			if compared == 0 || !sc.AllowOverwritingInlinedFields {
+				return nil, fmt.Errorf("(struct %s) duplicated key %s", t.String(), description.name)
+			}
+			if compared > 0 {
+				continue
+			}
 		}
 
 		sd.fm[description.name] = description
-		sd.fl = append(sd.fl, description)
 	}
 
+	// Add all the fieldDescriptions and sort them so the final document is correctly ordered
+	for _, desc := range sd.fm {
+		sd.fl = append(sd.fl, desc)
+	}
+	sort.Sort(byIndex(sd.fl))
+
 	sc.l.Lock()
 	sc.cache[t] = sd
 	sc.l.Unlock()
 
 	return sd, nil
 }
 
+// compareField compares two fieldDescriptions. Returns 1 if fd1 overwrites fd2, -1 if fd2 overwrites fd1, and 0 if
+// they are equally dominant
+func compareField(fd1, fd2 fieldDescription) int {
```

Comment:
I asked about this in Slack too, but I'll put it here so it doesn't get lost. Can you explain the rationale for this function? Why can we compare the length of each field's `inline` list to determine which field dominates the other? Also, can you add a comment explaining this so it's clearer for future readers?

Reply:
went away